usr/bin/cylc

```diff
@@ -16,54 +16,132 @@
+#   $ . ~/.bashrc
+#
+# Note you can also `pip install` cylc-flow releases under a common root
+# location for CYLC_VERSION based selection, like for Conda:
+#   $ CYLC_HOME_ROOT=$HOME/cylc-flow/releases  # (set in this wrapper too)
+#   $ python -m venv $CYLC_HOME_ROOT/cylc-flow-8.0a2
+#   $ . $CYLC_HOME_ROOT/cylc-flow-8.0a2/bin/activate
+#   $ python -m pip install cylc-flow==8.0a2
+# But there is no need to do this if you have full system releases - not just
+# cylc-flow - installed via Conda anyway.
  
-#_____________________________
-# NOTES:
-# * Users will typically accept the default or set $CYLC_VERSION, but
-#   they may set $CYLC_HOME to pick up a private cylc installation.
-# * Developers may set $CYLC_HOME to pick up their own repository clone.
-# * Running suites export $CYLC_VERSION to task environments, so that
-#    task messaging commands etc. pick up the compatible cylc version.
-# * Note that $CYLC_ROOT_HOME can be overridden in the environment too.
+##############################!!! EDIT ME !!!##################################
+# Central Cylc Conda releases:
+CYLC_HOME_ROOT="${CYLC_HOME_ROOT:-/opt}"
+# User Conda releases:
+CYLC_HOME_ROOT_ALT="${CYLC_HOME_ROOT_ALT:-$HOME/miniconda3/envs}"
 
-CYLC_HOME_ROOT=${CYLC_HOME_ROOT:-/opt} # !!! EDIT ME !!!
+# Legacy Cylc 7 locations if needed:
+CYLC7_HOME_ROOT="${CYLC7_HOME_ROOT_ALT:-/opt}" 
+CYLC7_HOME_ROOT_ALT="${CYLC7_HOME_ROOT_ALT:-$HOME/cylc/cylc-7}" 
+###############################################################################
 
-if [[ -z ${CYLC_HOME:-} ]]; then
-    if [[ -n ${CYLC_VERSION:-} && -d $CYLC_HOME_ROOT/cylc-$CYLC_VERSION ]]; then
-        CYLC_HOME=$CYLC_HOME_ROOT/cylc-$CYLC_VERSION
+if [[ -z "${CYLC_HOME:-}" ]]; then
+    # If CYLC_HOME is not defined find it with CYLC_HOME_ROOT and CYLC_VERSION.
+    if [[ -n ${CYLC_VERSION:-} ]]; then
+        # Look for matching version in root locations
+        for ROOT in "${CYLC_HOME_ROOT}" "${CYLC_HOME_ROOT_ALT}" \
+                "${CYLC7_HOME_ROOT}" "${CYLC7_HOME_ROOT_ALT}"; do
+            for NAME in "cylc" "cylc-flow"; do
+                if [[ -d "${ROOT}/${NAME}-${CYLC_VERSION}" ]]; then
+                    CYLC_HOME="${ROOT}/${NAME}-${CYLC_VERSION}"
+                    break 2
+                fi
+            done
+        done
+    else
+        echo 1>&2 "ERROR: CYLC_HOME not defined or found"
+        exit 1
+    fi
+fi
+if [[ -d "${CYLC_HOME}/conda-meta" ]]; then
+    # A Cylc 8 Conda environment
+    if [[ -f "${CYLC_HOME}/bin/activate" ]]; then
+        # A conda pack installation
+        . "${CYLC_HOME}/bin/activate" || exit 1
     else
-        CYLC_HOME=$CYLC_HOME_ROOT/cylc
+        # A normal conda environment
+        . "$(dirname "${CYLC_HOME}")"/../etc/profile.d/conda.sh
+        conda activate "$(basename "${CYLC_HOME}")" || exit 1
     fi
+elif [[ -f "${CYLC_HOME}/bin/activate" ]]; then
+    # A Cylc 8 Python venv
+    . "${CYLC_HOME}/bin/activate" || exit 1
```

Comment:
venv & conda pack use the same logic so this can be simplified.
Also, use bash parameter expansion to avoid the extra forks?
```bash
if [[ -f "${CYLC_HOME}/bin/activate" ]]; then
    # A Python venv or conda pack installation
    . "${CYLC_HOME}/bin/activate" || exit 1
elif [[ -d "${CYLC_HOME}/conda-meta" && -f "${CYLC_HOME%/*}/../etc/profile.d/conda.sh" ]]; then
    # A normal conda environment
    . "${CYLC_HOME%/*}/../etc/profile.d/conda.sh"
    conda activate "${CYLC_HOME##*/}" || exit 1
```

Reply:
Yep.

>  Also, use bash parameter expansion to avoid the extra forks?

Good idea, it might be slightly more efficient.  We can also do that with the final `basename` call in the `exec` line (which has been there forever in Cylc 7).